File: src/ladym/layers/episodic.py
"""L1 — Episodic memory.

Time-stamped events ``(agent, action, observation, outcome)`` — the hippocampal fast-write
store. Items here are *raw* and decay over time; consolidation promotes the salient ones to
L2 (semantic) and clusters of successful ones to L3 (procedural).
"""

from __future__ import annotations

from typing import Any

from ..schema import Layer, Memory, MemoryType
from ..storage.embeddings import EmbeddingProvider
from ..storage.store import SQLiteStore


class EpisodicMemory:
    def __init__(self, store: SQLiteStore, embedder: EmbeddingProvider,
                 workspace: str = "default"):
        self.store = store
        self.embedder = embedder
        self.workspace = workspace

    def record(self, *, agent: str, action: str, observation: str = "",
               outcome: str = "", tags: list[str] | None = None,
               metadata: dict[str, Any] | None = None) -> Memory:
        """Append an episodic event. ``content`` is a rendered sentence for embedding."""
        parts = [f"agent={agent}", f"action={action}"]
        if observation:
            parts.append(f"observation={observation}")
        if outcome:
            parts.append(f"outcome={outcome}")
        content = " | ".join(parts)
        meta = dict(metadata or {})
        meta.setdefault("agent", agent)
        meta.setdefault("action", action)
        if observation:
            meta["observation"] = observation
        if outcome:
            meta["outcome"] = outcome

        mem = Memory(
            layer=Layer.EPISODIC,
            type=MemoryType.EVENT,
            content=content,
            summary=f"{agent}: {action}",
            tags=tags or [],
            metadata=meta,
            source=agent,
            workspace=self.workspace,
        )
        vec = self.embedder.embed(content)
        self.store.put_memory(mem, vector=vec)
        return mem

    def recent(self, limit: int = 50) -> list[Memory]:
        rows = self.store.conn.execute(
            "SELECT * FROM memories WHERE layer = ? AND workspace = ? "
            "ORDER BY created_at DESC LIMIT ?",
            (Layer.EPISODIC.value, self.workspace, limit),
        ).fetchall()
        return [self.store._row_to_memory(r) for r in rows]

File: layers/episodic.go
package layers

import (
	"sort"
	"strings"

	"github.com/ProjAnvil/LadyM/schema"
	"github.com/ProjAnvil/LadyM/storage"
)

// EpisodicMemory is L1 — time-stamped events.
type EpisodicMemory struct {
	Store     *storage.SQLiteStore
	Embedder  storage.EmbeddingProvider
	Workspace string
}

// NewEpisodicMemory builds an EpisodicMemory.
func NewEpisodicMemory(store *storage.SQLiteStore, embedder storage.EmbeddingProvider, workspace string) *EpisodicMemory {
	if workspace == "" {
		workspace = "default"
	}
	return &EpisodicMemory{Store: store, Embedder: embedder, Workspace: workspace}
}

// Record appends an episodic event. Content is a rendered sentence for embedding.
func (e *EpisodicMemory) Record(agent, action, observation, outcome string, tags []string, metadata map[string]any) (*schema.Memory, error) {
	parts := []string{"agent=" + agent, "action=" + action}
	if observation != "" {
		parts = append(parts, "observation="+observation)
	}
	if outcome != "" {
		parts = append(parts, "outcome="+outcome)
	}
	content := strings.Join(parts, " | ")
	if metadata == nil {
		metadata = map[string]any{}
	}
	meta := map[string]any{}
	for k, v := range metadata {
		meta[k] = v
	}
	// setdefault semantics (Python): caller-supplied agent/action keys win.
	if _, ok := meta["agent"]; !ok {
		meta["agent"] = agent
	}
	if _, ok := meta["action"]; !ok {
		meta["action"] = action
	}
	if observation != "" {
		meta["observation"] = observation
	}
	if outcome != "" {
		meta["outcome"] = outcome
	}
	if tags == nil {
		tags = []string{}
	}

	m := schema.NewMemory(schema.LayerEpisodic, schema.TypeEvent)
	m.Content = content
	m.Summary = agent + ": " + action
	m.Tags = tags
	m.Metadata = meta
	m.Source = agent
	m.Workspace = e.Workspace

	vec, err := e.Embedder.Embed(content)
	if err != nil {
		return nil, err
	}
	if err := e.Store.PutMemory(m, vec); err != nil {
		return nil, err
	}
	return m, nil
}

// Recent returns the most recent episodes, newest first (created_at DESC).
// limit <= 0 returns all episodes in the workspace.
func (e *EpisodicMemory) Recent(limit int) ([]*schema.Memory, error) {
	mems, err := e.Store.IterMemories(e.Workspace, string(schema.LayerEpisodic), "")
	if err != nil {
		return nil, err
	}
	sort.SliceStable(mems, func(i, j int) bool { return mems[i].CreatedAt > mems[j].CreatedAt })
	if limit > 0 && len(mems) > limit {
		mems = mems[:limit]
	}
	return mems, nil
}
